spec/Protocol.md

```diff
@@ -56,93 +129,250 @@ including the newline character that marks the end of the message.
+- `leaves` (array of string message IDs) an array of recent message IDs.
+  - This array may have any number of elements (including none), but all elements must be string message IDs.
+  - This array should only include messages for which the `Server` has no known reply. This keeps the array smaller.
+  - The server may, at any time, shrink this array by whatever heuristic it desires. A recommended heuristic would
+    be to remove the `id`s of `Post`s that have not been replied to within 24 hours.
+- `major` (integer) the major protocol version number of the server. (For this specification, 0)
+- `minor` (integer) the minor protocol version number of the server. (For this specification, 2)
+- `meta`: Arbitrary metadata intended to facilitate protocol extensions.
+  - All keys must contain only characters valid in Base64.
+  - Maximum key length is 128 bytes.
+  - All values must be valid UTF-8 strings.
+  - Values may not contain control characters.
+  - Maximum value length is 1024 bytes.
+  - Maximum of 32 key-value pairs.
 
 A sample WELCOME message looks like this:
 
 ```json
-{"Type":0,"Root":"f4ae0b74-4025-4810-41d6-5148a513c580","Recent":["92d24e9d-12cc-4742-6aaf-ea781a6b09ec","880be029-0d7c-4a3f-558d-d90bf79cbc1d"],"Major":0,"Minor":1}
+{
+  "type": 0,
+  "root": "f4ae0b74-4025-4810-41d6-5148a513c580",
+  "leaves": [
+    "92d24e9d-12cc-4742-6aaf-ea781a6b09ec",
+    "880be029-0d7c-4a3f-558d-d90bf79cbc1d"
+  ],
+  "major": 0,
+  "minor": 2,
+  "meta": {
+    "key": "value"
+  }
+}
 ```
 
 #### QUERY
 
-QUERY messages are used by clients to request message information from a server.
+QUERY messages are used by clients to request post information from a server.
 
 QUERY messages contain the following JSON fields:
 
-- `Type` (integer) the message type, should be a 1 for QUERY
-- `UUID` (string message ID) the id of the message that is being queried
+- `type` (integer) the message type, should be a 1 for QUERY
+- `ids` (array of string post `id`s) the requested `Post` `id`s
+  - At most 1024 `Post` `id`s may be included in a single `QUERY`.
+- `meta`: Arbitrary metadata intended to facilitate protocol extensions.
+  - All keys must contain only characters valid in Base64.
+  - Maximum key length is 128 bytes.
+  - All values must be valid UTF-8 strings.
+  - Values may not contain control characters.
+  - Maximum value length is 1024 bytes.
+  - Maximum of 32 key-value pairs.
+
 
 A sample QUERY message looks like this:
 
 ```json
-{"Type":1,"UUID":"f4ae0b74-4025-4810-41d6-5148a513c580"}
+{
+  "type": 1,
+  "ids": [
+    "f4ae0b74-4025-4810-41d6-5148a513c580",
+    "3397ae9e-00ca-11e9-ac8d-57df8dd16d7f"
+  ],
+  "meta": {
+    "key": "value"
+  }
+}
+```
+
+#### NEW
+
+NEW messages are used by the server to deliver post information to clients. They are also
+used by the client to send new messages to the server. In contrast to the version 0.1
+protocol, NEW messages are not used to respond to QUERY messages.
```

Comment:
I suppose it helps with clarification for the client. I'm just curious about the motivation for this change, since clients are currently setup to receive new messages and query responses as the same message type.

Reply:
Having explicit query responses helps clients differentiate between things that are happening in real time and posts that were sent in the past without having to parse all of their timestamps.

Now that you've asked though, I'm second-guessing myself. What use-cases was I thinking of when I decided to have explicit query responses? I think maybe it was a clarity thing, but the advantage isn't very great. Perhaps we should think on this more before adding it.

Reply:
As I mentioned over the phone, I think this is a good idea, I was mostly curious about the motivation. I think it should be kept.